pvHelpers/proxy_config/pac_parser.py

```diff
@@ -0,0 +1,164 @@
+import pacparser
+import requests
+
+from ..misc import g_log
+
+
+class Pac(object):
+
+    def __init__(self, pac_url, proxy_auth=None):
+        pacparser.init()
+        # determine whether this is a url
+        # or a file path
+        downloaded_pac = download_pac([pac_url])
+        if downloaded_pac is not None:
+            g_log.debug("pac url is a url: {}".format(pac_url))
+            pacparser.parse_pac_string(downloaded_pac)
+        else:
+            g_log.debug("pac url is a local file: {}".format(pac_url))
+            pacparser.parse_pac_file(pac_url)
+
+        self._proxy_auth = proxy_auth
+
+    @property
+    def proxy_auth(self):
+        """Proxy authentication object."""
+        return self._proxy_auth
+
+    @proxy_auth.setter
+    def proxy_auth(self, value):
+        self._proxy_auth = value
+
+    def __getProxies(self, url, host=None):
+        """
+        :return: a string of proxies from FindProxyForURL() in pac file js
+        or 'None' if we can't process the url.
+        """
+        try:
+            proxy_str = pacparser.find_proxy(url, host)
+        except pacparser.URLError as e:
+            g_log.error(e)
+            return None
+        else:
+            return parse_pac_value(proxy_str, self._proxy_auth)
+
+    def getProxies(self, url, host=None):
+        """
+        :return: a list of proxy dict that can be passed to
+        requests' proxies parameter. Each proxy dict is meant to be used in a
+        fallover fashion. 'None' means the url requires no proxy.
+        """
+

```

Comment:
I think lines 13-20, should happen in `getProxies` if not fetched. `if (pac file not fetched): attempt fetch it`